probe-rs/src/probe/blackmagic/arm.rs

```diff
@@ -130,75 +147,144 @@ impl BlackMagicProbeArmDebug {
+            if ctrl_reg.orun_detect() != self.use_overrun_detect {
+                tracing::debug!("Setting orun_detect: {}", self.use_overrun_detect);
+                // only write if there’s a need for it.
+                ctrl_reg.set_orun_detect(self.use_overrun_detect);
+                self.write_dp_register(dp, ctrl_reg)?;
+            }
+
+            let idr: DebugPortId = self.read_dp_register::<DPIDR>(dp)?.into();
+            tracing::info!(
+                "Debug Port version: {} MinDP: {:?}",
+                idr.version,
+                idr.min_dp_support
+            );
+
+            let state = self
+                .dps
+                .get_mut(&dp)
+                .expect("This DP State was inserted earlier in this function");
+            state.debug_port_version = idr.version;
+            if idr.version == DebugPortVersion::DPv3 {
+                state.current_select = SelectCache::DPv3(SelectV3(0), Select1(0));
+            }
+        } else if switched_dp {
+            let sequence = self.sequence.clone();
+
+            let start_span = tracing::debug_span!("debug_port_start").entered();
+            sequence.debug_port_start(self, dp)?;
+            drop(start_span);
+        }
+
+        // note(unwrap): Entry gets inserted above
+        Ok(self.dps.get_mut(&dp).unwrap())
+    }
+
+    fn select_dp_and_dp_bank(
+        &mut self,
+        dp: DpAddress,
+        dp_register_address: &DpRegisterAddress,
+    ) -> Result<(), ArmError> {
+        let dp_state = self.select_dp(dp)?;
+
+        // DP register addresses are 4 bank bits, 4 address bits. Lowest 2 address bits are
+        // always 0, so this leaves only 4 possible addresses: 0x0, 0x4, 0x8, 0xC.
+        // On ADIv5, only address 0x4 is banked, the rest are don't care.
+        // On ADIv6, address 0x0 and 0x4 are banked, the rest are don't care.
+
+        let &DpRegisterAddress {
+            bank,
+            address: addr,
+        } = dp_register_address;
+
+        if addr != 0 && addr != 4 {
+            return Ok(());
+        }
+
+        let bank = bank.unwrap_or(0);
+
+        if bank != dp_state.current_select.dp_bank_sel() {
+            dp_state.current_select.set_dp_bank_sel(bank);
+
+            tracing::debug!("Changing DP_BANK_SEL to {:x?}", dp_state.current_select);
+
+            match dp_state.current_select {
+                SelectCache::DPv1(select) => self.write_dp_register(dp, select)?,
+                SelectCache::DPv3(select, _) => self.write_dp_register(dp, select)?,
+            }
+        }
+
         Ok(())
     }
+
+    fn select_ap(&mut self, ap: &FullyQualifiedApAddress) -> Result<(u8, u8), ArmError> {
+        let apsel = match ap.ap() {
+            crate::architecture::arm::ApAddress::V1(val) => *val,
+            crate::architecture::arm::ApAddress::V2(_) => {
+                return Err(ArmError::NotImplemented(
+                    "AP address v2 currently unsupported",
```

Comment:
It seems very odd that we can enumerate the APs, but not actually use them. What's the purpose of adding DPv3 support but not APv2? APv2+DPv3 is the ADIv6 spec, so one without the other feels a bit pointless.

Reply:
It's only partially implemented. Newer versions of the BMP protocol implement ADIv6 acceleration, so we will need to fall back if it's not present.

Unfortunately I don't have any RP2350 so I'm unable to work on implementing APv2.